File: python/protein-translation/protein_translation.py
"""
Calculate a list of amino acids for a given codon strand.
"""

codon_amino_acid = {
    'AUG': 'Methionine',
    'UUU': 'Phenylalanine',
    'UUC': 'Phenylalanine',
    'UUA': 'Leucine', 
    'UUG': 'Leucine',
    'UCU': 'Serine',
    'UCC': 'Serine',
    'UCA': 'Serine',
    'UCG': 'Serine',
    'UAU': 'Tyrosine',
    'UAC': 'Tyrosine',
    'UGU': 'Cysteine',
    'UGC': 'Cysteine',
    'UGG': 'Tryptophan',
    'UAA': '',
    'UAG': '',
    'UGA': '',
}

def proteins(strand):
    """
    For the given codon strand calculate the amino acids it produces.
    """
    ret = []

    for index in range(0, len(strand), 3):
        key = strand[index:index + 3]
        if key not in codon_amino_acid:
            break   # bad key = stop I decided
        amino_acid = codon_amino_acid[key]
        if len(amino_acid) == 0:
            break
        ret.append(amino_acid)

    return ret 
    

File: php/protein-translation/ProteinTranslation.php
<?php
// Protein Translation Exercise

declare(strict_types=1);

$GLOBALS['codonAminoAcid'] = array(
    "AUG" => "Methionine",
    "UUU" => "Phenylalanine",
    "UUC" => "Phenylalanine",
    "UUA" => "Leucine",
    "UUG" => "Leucine",
    "UCU" => "Serine",
    "UCC" => "Serine",
    "UCA" => "Serine",
    "UCG" => "Serine",
    "UAU" => "Tyrosine",
    "UAC" => "Tyrosine",
    "UGU" => "Cysteine",
    "UGC" => "Cysteine",
    "UGG" => "Tryptophan",
    "UAA" => "STOP",
    "UAG" => "STOP",
    "UGA" => "STOP",
);

class ProteinTranslation
{
    // From a given string of codon return an array of the amino acids produced. Stopping if you reach a stop codon.
    // @param $rna: codon list every three characters is a single codon. No spaces all one big string.
    // @returns: Array of Amino Acids that would be produced by the rna sequence.
    public function getProteins($rna)
    {
        global $codonAminoAcid;
        $result = array();        

        // If length is not divisible by three the last codon will be 
        // too short to be in the lookup array and you will get the 
        // Invalid argument exception.
        foreach(str_split($rna, 3) as $codon) {
            if (array_key_exists($codon, $codonAminoAcid)) {
                $aminoAcid = $codonAminoAcid[$codon];
                if ($aminoAcid == "STOP") {
                    break;
                }
                $result[] = $aminoAcid;
            } else {
                throw new InvalidArgumentException("Invalid codon: " . $codon);
            }
        }

        return $result;
    }
}
